### PR TITLE
ci: Comment with schema-diff output on Github PR

### PR DESCRIPTION
**Pull request by hugoamvieira**

### CODE REVIEW

File: .github/workflows/schema-diff.yml [hunk 1]
```diff
@@ -15,4 +15,49 @@ jobs:
...
+              echo "Unchanged"
+            fi
+          }
+
+          # Build heredoc with all the diffs the schema diff tool generates.
+          cat << EOF > github-comment.txt
+          ### Database schema diff between ${GITHUB_BASE_REF} and ${GITHUB_HEAD_REF}
+
+          #### Functions
+          $(getdiff funcs.diff)
```
Comment (tmessi) [inline .github/workflows/schema-diff.yml:38]: I think the actual file locations are:

```
.schema-diff/funcs.diff
.schema-diff/tables.diff
.schema-diff/views.diff
.schema-diff/triggers.diff
.schema-diff/indexes.diff
```

Comment (hugoamvieira) [reply .github/workflows/schema-diff.yml:38]: Oh yeah 🤦

**Comment by hugoamvieira** [comment]
A test comment

**Comment by hugoamvieira** [comment]
A different test comment

**Comment by hugoamvieira** [comment]
Database schema diff between

some comment

**Comment by hugoamvieira** [comment]
Database schema diff between

some test comment that shouldn't be hidden

File: scripts/schema-diff-hide-gh-comments.sh [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+# List all comments for the Pull Request we're working on.
+echo "Listing all issue comments for PR #$gh_pr_number"
+curl -fsSX GET \
+  -H "Accept: application/vnd.github+json" \
+  -H "Authorization: Bearer $gh_token" \
+  --output response.json \
+  $gh_api_url/repos/$gh_repo/issues/$gh_pr_number/comments
+if [ $? -ne 0 ]; then
+  die "Failed to list all issue comments for PR #$gh_pr_number"
+fi
```
Comment (tmessi) [inline scripts/schema-diff-hide-gh-comments.sh:38]:
```suggestion
curl -fsSX GET \
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer $gh_token" \
  --output response.json \
  $gh_api_url/repos/$gh_repo/issues/$gh_pr_number/comments || die "Failed to list all issue comments for PR #$gh_pr_number"

File: scripts/schema-diff-hide-gh-comments.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+# Hide Comments
+echo 'Issuing GraphQL calls to GitHub to hide previous schema-diff comments'
+while IFS= read -r hide_query; do
+  curl -fsSX POST \
+    -H "Authorization: Bearer $gh_token" \
+    -d "$hide_query" \
+    $gh_gql_url
+  if [ $? -ne 0 ]; then
+    die "Failed to issue request to minimize comment"
+  fi
```
Comment (tmessi) [inline scripts/schema-diff-hide-gh-comments.sh:87]:
```suggestion
  curl -fsSX POST \
    -H "Authorization: Bearer $gh_token" \
    -d "$hide_query" \
    $gh_gql_url || die "Failed to issue request to minimize comment"
```

File: scripts/schema-diff-hide-gh-comments.sh [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+while IFS= read -r node_id; do
+  echo "mutation { minimizeComment(input: {subjectId: $node_id, classifier: OUTDATED}) { minimizedComment { isMinimized } } }" >> graphql.txt
+done <<< "$(cat comment_ids.txt)"
+
+# Parse it through jq to build a valid json object.
+while IFS= read -r graphql; do
+  jq --null-input -c --arg q "$graphql" '{"query": $q}' >> hide_queries.json
+  if [ $? -ne 0 ]; then
+    die "Failed to create http minimizeComment query for graphql query $graphql"
+  fi
```
Comment (tmessi) [inline scripts/schema-diff-hide-gh-comments.sh:75]:
```suggestion
  jq --null-input -c --arg q "$graphql" '{"query": $q}' >> hide_queries.json || die "Failed to create http minimizeComment query for graphql query $graphql"
```

File: scripts/schema-diff-hide-gh-comments.sh [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  --arg user_login "$gh_user_login" \
+  --arg pfx "$gh_comment_prefix" \
+  '.[] |
+    select(.user.login == $user_login) |
+    select(.body | startswith($pfx)) |
+    .node_id
+  ' > comment_ids.txt
+if [ $? -ne 0 ]; then
+  die 'Failed to parse issue comments response'
+fi
```
Comment (tmessi) [inline scripts/schema-diff-hide-gh-comments.sh:57]:
```suggestion
cat response.json \
| jq \
  --arg user_login "$gh_user_login" \
  --arg pfx "$gh_comment_prefix" \
  '.[] |
    select(.user.login == $user_login) |
    select(.body | startswith($pfx)) |
    .node_id
  ' > comment_ids.txt || die 'Failed to parse issue comments response'
```

File: scripts/schema-diff-hide-gh-comments.sh [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  -H "Authorization: Bearer $gh_token" \
+  --output response.json \
+  $gh_api_url/repos/$gh_repo/issues/$gh_pr_number/comments
+if [ $? -ne 0 ]; then
+  die "Failed to list all issue comments for PR #$gh_pr_number"
+fi
+
+if [ $(jq 'length' response.json) -eq 0 ]; then
+  die "No comments found for PR #$gh_pr_number, nothing to hide"
+fi
```
Comment (tmessi) [inline scripts/schema-diff-hide-gh-comments.sh:42]:
```suggestion
[[ $(jq 'length' response.json) -gt 0 ]] || die "No comments found for PR #$gh_pr_number, nothing to hide"

File: scripts/schema-diff-hide-gh-comments.sh [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    select(.body | startswith($pfx)) |
+    .node_id
+  ' > comment_ids.txt
+if [ $? -ne 0 ]; then
+  die 'Failed to parse issue comments response'
+fi
+
+if [ $(cat comment_ids.txt | wc -l) -eq 0 ]; then
+  die "No comments matching message prefix and github user id ($gh_user_login) found"
+fi
```
Comment (tmessi) [inline scripts/schema-diff-hide-gh-comments.sh:61]:
```suggestion
[[ $(cat comment_ids.txt | wc -l) -gt 0 ]] || die "No comments matching message prefix and github user id ($gh_user_login) found"
```

**Review by tmessi — COMMENTED** [review]
Looks good, just some suggestions that are largely preference. When using a `die` function doing `|| die` I think reads more intuitively.

**Comment by hugoamvieira** [comment]
> Looks good, just some suggestions that are largely preference. When using a die function doing || die I think reads more intuitively.

Definitely reads better to me! I wasn't aware of this pattern, TIL! 

Q: When using `|| die`, is splitting the statements over new lines using `\` still valid? Since some of the error messages are a bit lengthy, I think it'd make sense to split.

Eg:
```bash
do_thing \
  with_arg1 \
  with_arg2 \
|| die "oops failed"
```

**Comment by tmessi** [comment]
> Q: When using `|| die`, is splitting the statements over new lines using `\` still valid? Since some of the error messages are a bit lengthy, I think it'd make sense to split.
> 
> Eg:
> 
> ```shell
> do_thing \
>   with_arg1 \
>   with_arg2 \
> || die "oops failed"
> ```

Yep

**Comment by hugoamvieira** [comment]
@tmessi All addressed with https://github.com/hashicorp/boundary/pull/3013/commits/d3e9e6f5f885f1cfd839af7e3372decfd3d51904 and https://github.com/hashicorp/boundary/pull/3013/commits/d3a0de38cd660c569663c9953ebd7376af1f4dd7

**Review by tmessi — APPROVED** [review]

**Pull request merged**